File: cloudflare/control-plane/src/artifact-manifest.ts
import { Json, requireThat, stable } from "./shared";

/** v3 changes only manifest representation, never prepared object bytes or order. */
export function compactGraph(graph: Json): Json {
  const { api_objects: objects, api_methods: descriptors, ...metadata } = graph;
  const positions = new Map<string, number>(objects.map((ref: Json, i: number) => [ref.sha256, i]));
  const methods: unknown[][] = [], scopes: string[] = [], entries: unknown[][] = [];
  const methodPositions = new Map<string, number>(), scopePositions = new Map<string, number>();
  for (const descriptor of descriptors) {
    const method = [descriptor.method, descriptor.method_version], key = stable(method);
    if (!methodPositions.has(key)) { methodPositions.set(key, methods.length); methods.push(method); }
    if (!scopePositions.has(descriptor.scope)) { scopePositions.set(descriptor.scope, scopes.length); scopes.push(descriptor.scope); }
    const index = descriptor.index ? positions.get(descriptor.index.sha256) : null;
    requireThat(index !== undefined && (!descriptor.index || objects[index!].bytes === descriptor.index.bytes), "invalid_prepared_reference");
    entries.push([methodPositions.get(key), scopePositions.get(descriptor.scope), descriptor.turn_id ?? null, index]);
  }
  return { ...metadata, api: { objects: objects.map((ref: Json) => [ref.sha256, ref.bytes]), methods, scopes, entries } };
}

export function expandGraph(graph: Json): Json {
  const { api, ...metadata } = graph;
  const objects = api.objects.map(([sha256, bytes]: [string, number]) => ({ kind: "api", sha256, bytes }));
  requireThat(new Set(objects.map((ref: Json) => ref.sha256)).size === objects.length, "invalid_prepared_reference");
  const at = (values: any[], position: number) => {
    requireThat(Number.isSafeInteger(position) && position >= 0 && position < values.length, "invalid_prepared_reference");
    return values[position];
  };
  const methods = api.entries.map(([method, scope, turn_id, index]: [number, number, string | null, number | null]) => {
    const [name, version] = at(api.methods, method);
    return { method: name, method_version: version, scope: at(api.scopes, scope), turn_id,
      index: index === null ? null : at(objects, index), error: index === null ? "remote_result_too_large" : null };
  });
  return { ...metadata, api_objects: objects, api_methods: methods };
}

export function expandManifest(manifest: Json): Json {
  return manifest.schema_version === "ct.artifact-manifest.v3"
    ? { ...manifest, schema_version: "ct.artifact-manifest.v2", graphs: manifest.graphs.map(expandGraph) }
    : manifest;
}

File: cloudflare/control-plane/src/artifact_manifest.py
"""Compact stored artifact-manifest representation."""

from __future__ import annotations

from typing import Any

from shared import require_that, stable

Json = dict[str, Any]


def compact_graph(graph: Json) -> Json:
    objects = graph["api_objects"]
    descriptors = graph["api_methods"]
    metadata = {
        key: value
        for key, value in graph.items()
        if key not in {"api_objects", "api_methods"}
    }
    positions = {reference["sha256"]: index for index, reference in enumerate(objects)}
    methods: list[list[Any]] = []
    scopes: list[str] = []
    entries: list[list[Any]] = []
    method_positions: dict[str, int] = {}
    scope_positions: dict[str, int] = {}
    for descriptor in descriptors:
        method = [descriptor["method"], descriptor["method_version"]]
        key = stable(method)
        if key not in method_positions:
            method_positions[key] = len(methods)
            methods.append(method)
        scope = descriptor["scope"]
        if scope not in scope_positions:
            scope_positions[scope] = len(scopes)
            scopes.append(scope)
        index = (
            positions.get(descriptor["index"]["sha256"])
            if descriptor.get("index")
            else None
        )
        if descriptor.get("index"):
            require_that(
                index is not None
                and objects[index]["bytes"] == descriptor["index"]["bytes"],
                "invalid_prepared_reference",
            )
        entries.append(
            [
                method_positions[key],
                scope_positions[scope],
                descriptor.get("turn_id"),
                index,
            ]
        )
    return {
        **metadata,
        "api": {
            "objects": [
                [reference["sha256"], reference["bytes"]] for reference in objects
            ],
            "methods": methods,
            "scopes": scopes,
            "entries": entries,
        },
    }


def expand_graph(graph: Json) -> Json:
    api = graph["api"]
    metadata = {key: value for key, value in graph.items() if key != "api"}
    objects = [
        {"kind": "api", "sha256": sha256, "bytes": byte_count}
        for sha256, byte_count in api["objects"]
    ]
    require_that(
        len({reference["sha256"] for reference in objects}) == len(objects),
        "invalid_prepared_reference",
    )

    def at(values: list[Any], position: Any) -> Any:
        require_that(
            isinstance(position, int)
            and not isinstance(position, bool)
            and 0 <= position < len(values),
            "invalid_prepared_reference",
        )
        return values[position]

    methods = []
    for method_position, scope_position, turn_id, index in api["entries"]:
        name, version = at(api["methods"], method_position)
        methods.append(
            {
                "method": name,
                "method_version": version,
                "scope": at(api["scopes"], scope_position),
                "turn_id": turn_id,
                "index": None if index is None else at(objects, index),
                "error": "remote_result_too_large" if index is None else None,
            }
        )
    return {**metadata, "api_objects": objects, "api_methods": methods}


def expand_manifest(manifest: Json) -> Json:
    if manifest.get("schema_version") != "ct.artifact-manifest.v3":
        return manifest
    return {
        **manifest,
        "schema_version": "ct.artifact-manifest.v2",
        "graphs": [expand_graph(graph) for graph in manifest["graphs"]],
    }
